### ISSUE TITLE
(5) UV Performance Issue  

### ISSUE THREAD

**Issue by kellyChex** [UCSCLibrary/dams_project_mgmt#36]
### Descriptive Summary

Rachel and I also noticed that the UV will not always display the images. 

### Actual Behavior

When we click to view via UV we see a blank page instead of the image in the UV. 

![Screen Shot 2019-10-08 at 3 47 34 PM](https://user-images.githubusercontent.com/9621070/66438954-08714600-e9e3-11e9-8a62-98cfa04615ce.png)

Additionally, it takes a long while for some of the images to load. 

### Solutions to Implement:
- [x] 1. Cache the manifest server side 
- [x] 2. Work with UCSC (Daniel) to give more resources to our server  
- [ ] 3. Optimize the process in Fedora
- [x] 4. Have Rails make the image server pre cache the images it will need instead of waiting until user requests to see them.

**Comment by NedHenry** [comment]
I did a bit of investigating into this. I loaded four works at random from each collection on the production server and a dozen works from the staging server (including the one pictured above). 

- I could not reproduce the bug reported above that results in a blank screen instead of loading the UV. Every page I loaded that was supposed to display the UV did indeed display the UV. If this bug is observed again, please take careful note of the circumstances under which you see it. I have a theory about what caused it (explained below), but I cannot verify that theory without reproducing the problem.

- I did observe slow loading times under certain circumstances, and I have a good theory about why that is and how to fix it. I think we are mainly running into performance issues with our image server, and we need to optimize how that caches images. Caching manifests (as suggested by Julie) will also help a lot with load times. Below I'm including more details about my observations and theories.

I think there are three overlapping issues going on. First, when we load an image in the UV for the first time ever, the UV has to dynamically create derivative files on the fly for whatever view we request. This happens the first time *anyone* looks at the image, then the derivative file is cached on the server side (which means for each image it doesn't happen once for each user, it happens once for everybody). This process makes the first load of an image a lot slower than the others. When we are doing quality testing on newly ingested works, we usually see the first load, which seems slower. We could address this either by configuring the image server to create these derivative files automatically on ingest in a size we think we'd like, or by making sure our image server always has enough resources to dynamically create the derivatives quickly enough (it currently doesn't do so badly for single images, but we also don't have much traffic yet).

So far, I mentioned that this issue is not so bad when we're dealing with single images - not enough to explain the performance problems in this bug report. In the production site, where there are only single images, loading times are pretty decent even for the first view of each image (which I why I hadn't seen the problem before).

The bigger problem comes when loading a work with many attached images. The image server seems to reloading the manifests and re-creating the thumbnails for each child work whenever the parent work loads, which slows down the load time for complex objects by a lot. We should find a way to pre-cache both the child work iiif manifests (Julie said she would help us with that) and pre-cache their thumbnail-sized derivatives on the image server (or even customize UV to link to static thumbnails for that panel pre-generated in Hyrax, if necessary).

Although improved caching will probably fix this problem, the severity

**Comment by NedHenry** [comment]
It is also worth noting that currently the production site has more computing resources devoted to it, both at the web application and image server level. Everything is slower in staging. We still need to address performance issues in staging, which will usually also occur in production once we have heavier load.

**Comment by NedHenry** [comment]
Update on recent work:
There were several separate problems that caused this issue in conjunction. Each of them is being addressed separately.
 - 1. Caching IIIF manifest files for each work
The universal viewer uses a work's IIIF manifest to determine what metadata and child work thumbnails to display. Hyrax is really slow at generating these manifests for a work with many child works. We implemented a cache (currently a filesystem cache, ideally a redis cache in the future) where we can pre-generate and store that manifest for each work. That can save as much as 6-10 seconds of load time when loading a complex object in the universal viewer. 

- 2. Caching thumbnails images on the image server
I made a change to the "create_derivatives" method for FileSets, causing them to ask the image server for various thumbnail files when the fileset is created. This should cause the most commonly used thumbnails to exist already when they are needed.
I also noticed that when the Universal Viewer loaded thumbnails, Cantaloupe wasn't using these cached thumbnails (though it's cache was working fine when I was testing it). I found that for some reason, UV was including a query string with a timestamp (e.g. `?t=123904839`) which confusingly prevented our image server from using its derivative cache appropriately (I don't understand why). As a result, whenever a complex object was loaded in UV, Cantaloupe got 40 or 50 simultaneous requests for thumbnails that it was creating on the fly. Some of the TIFs it was working with are 200MB in size, and this is too much to ask of an image server.
**Reverse Proxy Server**
I found that the easiest way to address the problem of accessing the derivative cache was to set up an Apache reverse proxy server in front of our image server. I was considering doing this anyway for https security/consistency, but it wasn't worth it until now. This layer gives the benefit of being able to strip off the query string before sending the request to the image server, so it always uses the derivative cache when appropriate. This also allows us more flexible security on our image server, which is a nice bonus. I have confirmed that this change allows complex objects to load much faster 
  - 3. Improving Image Server Peformance
